### ISSUE TITLE
Allow order to be preserved for messages while using multiple queue threads

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#3410]
When higher throughput is needed on a destination queue, multiple queue threads can be used. This comes with the caveat where message order is not guaranteed. However, there are times when you don't care about message order in general, _except_ for certain messages.

For example, if the destination is posting HL7 messages across many different facilities for many different patients, in general you don't care what order they post. However for messages pertaining to the _same_ patient, order **does** matter. You wouldn't want an A08 to post before the initial A04 registration, for instance.

In these cases, it would be useful to allow increasing message throughput while still keeping the order integrity to a degree.

Imported Issue. Original Details:
Jira Issue Key: MIRTH-3513
Reporter: narupley
Created: 2014-11-07T15:52:54.000-0800

**Comment by rbeckman-nextgen** [comment]
Revision 7453: Added the ability to preserve order among destination queue messages while still using multiple queue threads. The way this works is with a Thread Assignment Variable. When acquiring a message from the queue, the variable (e.g. with the key "mrn") is retrieved from the standard map lookup sequence. That value is then hashed and reduced via modulo to one of the queue threads. Because the hash function used (MurmurHash3 32-bit) is deterministic, the same input (e.g. "12345") will always produce the same hash value, and therefore the same queue thread. Only the queue thread that has been calculated for that connector message will be allowed to acquire the message; other threads will skip over it in the buffer.

Imported Comment. Original Details:
Author: narupley
Created: 2014-11-07T16:05:56.000-0800

**Comment by rbeckman-nextgen** [comment]
To test this issue, I performed the following procedure as outlined below.

Create Channels
I created 2 channels, the first being a simple Channel Reader and the second a JS Reader. I added a deploy script to the JS Reader channel so as to set and initialize a counter to 0, placing this in the global map. For the JS Script, I read out the variable, incremented it, and put it back into the map. I used this count as the content for my message. On the second channel, the Channel Reader, I made a simple transformer to essentially log out the message content.

Perform testing
I first verified that messages were able to be sent out of order. To test this, I enabled multiple queue threads on the JS Reader and enabled the destination queue. I stopped my Channel Reader and then let the JS Reader attempt to send messages to this destination. Once I got about 100k messages, I started the Channel Reader and let the messages process. I then took my logs and converted it into a csv file by replacing the spaces. I imported this file into a spreadsheet and then subtracted the values of the message content. When the differential was more than 1, the message was deemed to be out of order. I verified this happened throughout many sections within these log files using multiple threads with no Assignment Variable.

I then reconfigured the channel to now include a Thread Assignment Variable. I also modified my script to populate the assignment variable in a number of ways. The first time I made it so that every 5 messages populated the assignment variable with the same data. I then performed similar analysis as above to ensure that none of these messages appeared out of order. I repeated this process assigning every 2 messages to their own thread, and after performing this test, I tried as many possible combinations I could think of. Each time I performed analysis, the messages appeared to maintain their order.

Imported Comment. Original Details:
Author: jaysenp
Created: 2015-01-27T16:46:08.000-0800
